packages/contracts/contracts/deploy/PayableDAppFactory.sol

```diff
@@ -0,0 +1,144 @@
+        IERC20 _token,
+        IConsensus _consensus,
+        uint256 _price
+    ) {
+        factory = _factory;
+        token = _token;
+        consensus = _consensus;
+        price = _price;
+
+        // transfer ownership to the give owner
+        transferOwnership(_owner);
+    }
+
+    /// @dev This function deploys a new application, and registers it with the payment system
+    function newApplication(
+        address _dappOwner,
+        bytes32 _templateHash,
+        string memory _cid,
+        uint256 _runway
+    ) external override returns (CartesiDApp) {
+        // delegate call to factory
+        CartesiDApp dapp = factory.newApplication(
+            consensus,
+            _dappOwner,
+            _templateHash
+        );
+
+        // emit event to bind CID to application
+        emit MachineLocation(address(dapp), _cid);
+
+        // set initial runway
+        _extendRunway(msg.sender, dapp, _runway);
+
+        // return application
+        return dapp;
+    }
+
+    function importApplication(
+        CartesiDApp _dapp,
+        string memory _cid,
+        uint256 _runway
+    ) external {
+        require(
+            _dapp.getConsensus() == this.consensus(),
+            "PaymentDAppFactory: cannot import dapp with a different consensus"
+        );
+
+        require(
+            _dapp.owner() == msg.sender,
+            "PaymentDAppFactory: only owner can import application"
+        );
+
+        // emit location event, as DApp was not intantiated by this factory
+        emit MachineLocation(address(_dapp), _cid);
+
+        // extend runway
+        _extendRunway(msg.sender, _dapp, _runway);
+    }
+
+    function cost(uint256 _time) external view returns (uint256) {
+        return _time * price;
+    }
+
+    function extendRunway(
+        ICartesiDApp _dapp,
+        uint256 _time
+    ) external returns (uint256) {
+        return _extendRunway(msg.sender, _dapp, _time);
+    }
+
+    function _extendRunway(
+        address _msgSender,
+        ICartesiDApp _dapp,
+        uint256 _time
+    ) private returns (uint256) {
+        // calculate the token cost of the runway time
+        uint256 _cost = this.cost(_time);
+
+        // transfer tokens from sender to this contract
+        require(
+            token.transferFrom(_msgSender, address(this), _cost),
+            "PaymentDAppFactory: failed to transfer tokens"
+        );
+
+        uint256 currentRunway = runway[_dapp];
+
+        // if runway was in the past, update to now
+        currentRunway = currentRunway < block.timestamp
+            ? block.timestamp
+            : currentRunway;
+
+        // update dapp runway
+        uint256 newRunway = currentRunway + _time;
+        runway[_dapp] = newRunway;
+
+        // emit event so offchain gets notified
+        emit FinancialRunway(address(_dapp), newRunway);
+
+        return newRunway;
+    }
+
+    function withdraw(uint256 amount) external onlyOwner {
```

Comment:
Have you thought about making it `withdraw()` without an amount and always withdraw the max (`balanceOf(this)`)? I think its a more elegant way to do it, it's usually what you're going for when withdrawing. And makes the order of this transaction less important in the scenario in which someone deposited money right before you're `balanceOf` check and your `withdraw(balancOf)` call...which is not much, but its something

Reply:
Do you think we need both methods? Or just the `withdraw()`?
The simpler the better IMO.

Reply:
Agreed! I think just `withdraw()` is better. 
Also saves you from having to count the 18 zeros for decimals haha